lib/Sema/CSDiag.cpp

```diff
@@ -4361,8 +4361,13 @@ bool TypeChecker::diagnoseSelfAssignment(const Expr *E) {
   if (!AE)
     return false;
 
+  auto src = AE->getSrc();
+  auto loadSrc = dyn_cast<LoadExpr>(src);
+  if (dyn_cast<ParenExpr>(loadSrc->getSubExpr()))
```

Comment:
It would also be great to leave some comments here explaining the behavior change.  It's not obvious why a parenthesized rvalue gets to invoke side effects and a regular assignment doesn't.